domain/src/main/java/org/oppia/domain/analytics/EventStorageHelper.kt

```diff
@@ -0,0 +1,97 @@
+package org.oppia.domain.analytics
+
+import androidx.lifecycle.LiveData
+import org.oppia.app.model.EventLog
+import org.oppia.app.model.EventLogs
+import org.oppia.data.persistence.PersistentCacheStore
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.data.DataProviders
+import org.oppia.util.logging.Logger
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class EventStorageHelper @Inject constructor(
+  private val cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val logger: Logger
+) {
+  private val eventLogStore =
+    cacheStoreFactory.create("event_logs", EventLogs.getDefaultInstance())
+
+  fun addEventLog(eventLog: EventLog) {
+    eventLogStore.storeDataAsync(updateInMemoryCache = true) {
+      it.toBuilder().addEventLog(eventLog).build()
+    }.invokeOnCompletion {
+      it?.let {
+        logger.e(
+          "DOMAIN",
+          "Failed to store event log",
+          it
+        )
+      }
+    }
+  }
+
+  suspend fun getOptionalEvents(): MutableList<EventLog>{
+    var eventLogList: MutableList<EventLog> = mutableListOf()
+    val eventList = eventLogStore.readDataAsync().await().eventLogList
+    for (eventLog in eventList){
+      if(eventLog.priority == EventLog.Priority.OPTIONAL){
+        eventLogList.add(eventLog)
+      }
+    }
+    return eventLogList
+  }
+
+  suspend fun getLeastRecentEvent(): EventLog?{
+    var timestamp: Long = 0
+    var event: EventLog? = null
+    val eventList = eventLogStore.readDataAsync().await().eventLogList
+    for (eventLog in eventList){
+      if(eventLog.timestamp > timestamp){
+        timestamp = eventLog.timestamp
+        event = eventLog
+      }
+    }
+    return event
+  }
+
+  suspend fun getLeastRecentOptionalEvent(): EventLog?{
```

Comment:
I am wondering if we should merge this and the above function. As far as I understand they both will be used primarily to figure out which event to delete right? 
So maybe just getLeastRecentEvent which does the following:

- If there are optional events available, returns the least recent optional event, else
- Returns the least recent essential event.

Does that make sense? That way the logic is abstracted away and callers of the function don't need to think too much about it (But make the comment very clear)

This totally depends on how you are planning to use the storage helper:
If these two functions are meant to be used publicly, then abstract it . Callers shouldn't have to think about optional/essential events while trying to get rid of old events.
if these are helper functions to be used within the storage helper itself , then it makes sense to keep them separate

Reply:
I am planning to use it in the following way - 
- The interaction outside this class is purely done by the addEvent and the getAllEvents functions. 
- Rest of the functions remain helper functions for internal usage -- book keeping 

I am planning to do this this way because when these functions are called from the controller, we either have to save them (no connectivity) or get all the event logs (in case of presence of network). I don't think that there is any need for the controller to call specific events (optional or leastRecent). 

Does that help?

Reply:
Sounds good